### PR TITLE
Allow CREATE MATERIALIZED VIEW x TO y, allow ATTACH TABLE x

### PR DESCRIPTION
**Pull request by vavrusa**
This adds two conveniences:

* Allow creating materialized views without the inner table, using an already existing table instead. This is useful for detaching view while keeping underlying table readable. At least it makes it more clear in scenario with multiple replicas of the view. Previously it wasn't obvious that the `.inner.` table remains readable while the view is detached, and whether the view or the `.inner.` table should be replicated.
* Allow `ATTACH TABLE x` without defining table (or view) structure in cases where the table has been previously detached and the server knows the structure already from `SHOW CREATE TABLE`.

### CODE REVIEW

**Comment by vavrusa** [comment]
```
$ /usr/local/bin/python2 ../dbms/tests/clickhouse-test -b "$(pwd)/clickhouse-server" -c "$(pwd)/clickhouse-client" --no-zookeeper --no-shard -q $(pwd)/../dbms/tests/queries "(material|creat|attach)"

Running stateless tests.

00508_materialized_view_to:                                            [ OK ]
00472_create_view_if_not_exists:                                       [ OK ]
00258_materializing_tuples:                                            [ OK ]
00235_create_temporary_table_as:                                       [ OK ]
00180_attach_materialized_view:                                        [ OK ]
00101_materialized_views_and_insert_without_explicit_database:         [ OK ]
00083_create_merge_tree_zookeeper:                                     [ SKIPPED ] - no zookeeper
00014_1_create_table_with_nested:                                      [ OK ]
00013_1_create_table_with_arrays:                                      [ OK ]
```

File: dbms/src/Databases/DatabasesCommon.cpp [hunk 1]
```diff
@@ -58,6 +63,8 @@ std::pair<String, StoragePtr> createTableFromDefinition(
     /// We do not directly use `InterpreterCreateQuery::execute`, because
     /// - the database has not been created yet;
     /// - the code is simpler, since the query is already brought to a suitable form.
+    if (!ast_create_query.columns)
+        throw Exception("Missing definition of columns");
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabasesCommon.cpp]: Missing error code.

File: dbms/src/Databases/DatabasesCommon.cpp [hunk 2]
```diff
@@ -68,7 +75,11 @@ std::pair<String, StoragePtr> createTableFromDefinition(
     else if (ast_create_query.is_materialized_view)
         storage_name = "MaterializedView";
     else
+    {
+        if (!ast_create_query.storage)
+            throw Exception("Missing ENGINE definition");
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabasesCommon.cpp]: .

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 3]
```diff
@@ -483,6 +483,15 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
     String data_path = path + "data/" + database_name_escaped + "/";
     String metadata_path = path + "metadata/" + database_name_escaped + "/" + table_name_escaped + ".sql";
 
+    // If this is a stub ATTACH query, read the query definition from the database
+    if (create.attach && !create.storage && !create.columns)
+    {
+        auto query = context.getCreateQuery(database_name, table_name);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: Should mention in comment that `getCreateQuery` works even for detached tables.

**Pull request merged**